irmaclient/keyshare.go

```diff
@@ -81,29 +82,27 @@ func newKeyshareServer(schemeManagerIdentifier irma.SchemeManagerIdentifier) (*k
 	return ks, nil
 }
 
-func (ks *keyshareServer) HashedPin(pin string) string {
-	hash := sha256.Sum256(append(ks.Nonce, []byte(pin)...))
+func (kss *keyshareServer) HashedPin(pin string) string {
+	hash := sha256.Sum256(append(kss.Nonce, []byte(pin)...))
 	// We must be compatible with the old Android app here,
 	// which uses Base64.encodeToString(hash, Base64.DEFAULT),
 	// which appends a newline.
 	return base64.StdEncoding.EncodeToString(hash[:]) + "\n"
 }
 
-// startKeyshareSession starts and completes the entire keyshare protocol with all involved keyshare servers
+// newKeyshareSession starts and completes the entire keyshare protocol with all involved keyshare servers
 // for the specified session, merging the keyshare proofs into the specified ProofBuilder's.
 // The user's pin is retrieved using the KeysharePinRequestor, repeatedly, until either it is correct; or the
 // user cancels; or one of the keyshare servers blocks us.
 // Error, blocked or success of the keyshare session is reported back to the keyshareSessionHandler.
-func startKeyshareSession(
+func newKeyshareSession(
 	sessionHandler keyshareSessionHandler,
 	client *Client,
 	pin KeysharePinRequestor,
-	builders gabi.ProofBuilderList,
 	session irma.SessionRequest,
 	implicitDisclosure [][]*irma.AttributeIdentifier,
-	issuerProofNonce *big.Int,
-	timestamp *atum.Timestamp,
-) {
+	protocolVersion *irma.ProtocolVersion,
+) (*keyshareSession, bool) {
```

Comment:
It's still passed to the `keyshareSession` struct, only not here anymore but a few lines below. The reason for this has not changed: in case of attribute-based signatures sessions, the timestamp that is to be included in the attribute-based signature is an input parameter for the challenge when it is constructed as the hash over several parameters, [here](https://github.com/privacybydesign/irmago/blob/f2274128d6c5d0dd772f6a5180e44b22a1a681b2/irmaclient/keyshare.go#L422). This is part of the cryptographic mechanism that convinces the verifier that all disclosures and signatures that it receives come from a single wallet within a single session.

The reason that this timestamp was first passed as parameter to `startKeyshareSession` and now is set some lines below is because the order of some things has had to change a little:
* To construct the proof builders at the start of a session, I now need some keyshare session state in order to be able to contact the new endpoint `/getPs` in case of issuance sessions. That is why the `keyshareSession`, i.e. the output of `startKeyshareSession()`/`newKeyshareSession()`, is now passed to `session.getBuilders()`.
* Only after `session.getBuilders()` has completed has the timestamp been constructed, so only then can we put it on the `keyshareSession`.